Sources/X509/Extension Types/NameConstraints.swift

```diff
@@ -24,43 +24,266 @@ import SwiftASN1
+        
+        @inlinable
+        public var description: String {
+            "[\(self.joined(separator: ", "))]"
+        }
+        
+        @usableFromInline
+        var subtrees: [GeneralName]
+        
+        @inlinable
+        init(subtrees: [GeneralName]) {
+            self.subtrees = subtrees
+        }
+        
+        @inlinable
+        public init(_ elements: some Sequence<String>) {
+            self.subtrees = elements.map { .uniformResourceIdentifier($0) }
+        }
+        
+        @inlinable
+        public init(arrayLiteral elements: String...) {
+            self.init(elements)
+        }
+        
+        @inlinable
+        public func makeIterator() -> some IteratorProtocol<String> {
+            self.subtrees.lazy.compactMap {
+                if case .uniformResourceIdentifier(let name) = $0 {
+                    return name
+                } else {
+                    return nil
+                }
+            }.makeIterator()
+        }
+        
+        @inlinable
+        var filtered: some Sequence<GeneralName> {
+            self.subtrees.lazy.filter {
+                guard case .uniformResourceIdentifier = $0 else {
+                    return false
+                }
+                return true
+            }
+        }
+    }
+    
+    
+    /// The DNS name trees that are permitted in certificates issued by this CA.
     ///
     /// These restrictions are expressed in forms like `host.example.com`. Any DNS name that can be
     /// constructed by adding zero or more labels to the left-hand side of the name satifies the constraint.
-    public var excludedDNSDomains: [String] {
+    public var permittedDNSDomains: DNSNames {
         get {
-            self.excludedSubtrees.compactMap {
-                if case .dnsName(let name) = $0 {
-                    return name
-                } else {
-                    return nil
+            DNSNames(subtrees: permittedSubtrees)
+        }
+        set {
+            permittedSubtrees.removeAll {
+                guard case .dnsName = $0 else {
+                    return false
                 }
+                return true
             }
+            permittedSubtrees.append(contentsOf: newValue.filtered)
+        }
+    }
+
+    /// The DNS name trees that are forbidden in certificates issued by this CA.
+    ///
+    /// These restrictions are expressed in forms like `host.example.com`. Any DNS name that can be
+    /// constructed by adding zero or more labels to the left-hand side of the name satifies the constraint.
+    public var excludedDNSDomains: DNSNames {
+        get {
+            DNSNames(subtrees: excludedSubtrees)
         }
         set {
-            fatalError("TODO")
+            excludedSubtrees.removeAll {
+                guard case .dnsName = $0 else {
+                    return false
+                }
+                return true
+            }
+            excludedSubtrees.append(contentsOf: newValue.filtered)
```

Comment:
Hmm, these are fairly expensive. I wonder if we can just decide we don't need them.

Reply:
I don't think this is too expensive. Why do you think so? `removeAll(where:)` has linear time complexity as well as `append(contentsOf:)`. Any idea how to make it cheaper?

Reply:
Nope, or I would have suggested one ;).

The awkward reality is that a linear time property setter is rarely what you actually want, which is why I'm asking whether we can safely just remove them altogether.

Reply:
I have made them internal for now because it makes testing quite nice. If we have a real use case we can think about it again.